base/sort.jl

```diff
@@ -1757,6 +1759,41 @@ julia> v
 """
 sort(v::AbstractVector; kws...) = sort!(copymutable(v); kws...)
 
+function sort(x::NTuple;
+              alg::Algorithm=defalg(x),
+              lt=isless,
+              by=identity,
+              rev::Union{Bool,Nothing}=nothing,
+              order::Ordering=Forward,
+              scratch::Union{Vector, Nothing}=nothing)
+    # Can't do this check with type parameters because of https://github.com/JuliaLang/julia/issues/56698
+    scratch === nothing || eltype(x) == eltype(scratch) || throw(ArgumentError("scratch has the wrong eltype"))
+    _sort(x, alg, ord(lt,by,rev,order), (;scratch))::typeof(x)
+end
+# Folks who want to hack internals can define a new _sort(x::NTuple, ::TheirAlg, o::Ordering)
+# or _sort(x::NTuple{N, TheirType}, ::DefaultStable, o::Ordering) where N
+function _sort(x::NTuple, a::Union{DefaultStable, DefaultUnstable}, o::Ordering, kw)
+    # The unrolled tuple sort is prohibitively slow to compile for length > 9.
+    # See https://github.com/JuliaLang/julia/pull/46104#issuecomment-1435688502 for benchmarks
+    if length(x) > 9
+        v = copymutable(x)
```

Comment:
`Memory` should be slightly cheaper than `Vector`

```suggestion
        v = Memory{eltype(x)}(undef, length(x))
        copyto!(v, x)
```

Reply:
Interestingly, despite saving on allocation count, I'm benchmarking this proposal as a slight runtime regression:

```julia
julia> @eval Base.Sort function _sort(x::NTuple, a::Union{DefaultStable, DefaultUnstable}, o::Ordering, kw)
           # The unrolled tuple sort is prohibitively slow to compile for length > 9.
           # See https://github.com/JuliaLang/julia/pull/46104#issuecomment-1435688502 for benchmarks
           if length(x) > 9
               v = copymutable(x)
               _sort!(v, a, o, kw)
               typeof(x)(v)
           else
               _mergesort(x, o)
           end
       end
_sort (generic function with 6 methods)

julia> pr = 1e9*[(@b rand(NTuple{N, Int}) sort seconds=.3).time for N in 10:20]
11-element Vector{Float64}:
 23.248226950354614
 27.354570637119117
 28.32774390243903
 32.282442748091604
 34.18737672583827
 42.12754158964881
 42.25352112676057
 53.95136778115502
 52.37461300309597
 58.02180685358256
 62.82965299684544

julia> @eval Base.Sort function _sort(x::NTuple, a::Union{DefaultStable, DefaultUnstable}, o::Ordering, kw)
           # The unrolled tuple sort is prohibitively slow to compile for length > 9.
           # See https://github.com/JuliaLang/julia/pull/46104#issuecomment-1435688502 for benchmarks
           if length(x) > 9
               #v = copymutable(x)
               v = Memory{eltype(x)}(undef, length(x))
               copyto!(v, x)
               _sort!(v, a, o, kw)
               typeof(x)(v)
           else
               _mergesort(x, o)
           end
       end
_sort (generic function with 7 methods)

julia> proposed = 1e9*[(@b rand(NTuple{N, Int}) sort seconds=.3).time for N in 10:20]
11-element Vector{Float64}:
 23.86934673366834
 25.521250000000002
 29.057391304347828
 35.642857142857146
 40.700421940928265
 42.10526315789474
 45.05417607223476
 48.62700228832952
 56.9277108433735
 62.28717948717949
 63.79821958456973

julia> proposed ./ pr
11-element Vector{Float64}:
 1.0267168668234397
 0.9329793670886075
 1.025757342498637
 1.104094179643955
 1.1905102361997708
 0.9994711670580952
 1.0662821670428893
 0.9013117607245302
 1.086933298008491
 1.0735132679401136
 1.0154157558020083

julia> using Statistics]

julia> Statistics.quantile(proposed ./ pr, [0,.25,.5,.75,1])
5-element Vector{Float64}:
 0.9013117607245302
 1.0074434614300518
 1.0267168668234397
 1.0802232829743024
 1.1905102361997708

julia> mean(prop
propertynames
proposed
julia> mean(proposed ./ pr)
1.0384532189845943
```

I'm happy to include clear-cut performance improvements in this PR, but for things more tricky like this that may need to be coupled with changes elsewhere (like whatever is causing `Vector`s to outperform `Memory`s) I'd rather see them in a followup.